storage/src/nodestore/alloc.rs

```diff
@@ -147,12 +148,158 @@ pub fn area_size_to_index(n: u64) -> Result<AreaIndex, Error> {
-/// to be zero. This reserved zero can be used as a [None] value for some use cases. In particular,
-/// branches can use `Option<LinearAddress>` which is the same size as a [`LinearAddress`]
-pub type LinearAddress = NonZeroU64;
+// pub type LinearAddress = NonZeroU64;
+/// A linear address in the nodestore storage.
+///
+/// This represents a non-zero address in the linear storage space.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct LinearAddress(NonZeroU64);
+
+#[expect(unsafe_code)]
+unsafe impl bytemuck::ZeroableInOption for LinearAddress {}
+#[expect(unsafe_code)]
+unsafe impl bytemuck::PodInOption for LinearAddress {}
+
+impl LinearAddress {
+    /// Create a new `LinearAddress`, returns None if value is zero.
+    #[expect(clippy::missing_panics_doc)]
+    #[inline]
+    #[must_use]
+    pub const fn new(addr: u64) -> Option<Self> {
+        match addr {
+            0 => None,
+            _ => Some(LinearAddress(NonZeroU64::new(addr).expect("non zero"))),
+        }
+    }
+
+    /// Get the underlying address as u64.
+    #[inline]
+    #[must_use]
+    pub const fn get(&self) -> u64 {
+        self.0.get()
+    }
+
+    /// Check if the address is 8-byte aligned.
+    #[inline]
+    #[must_use]
+    pub const fn is_aligned(&self) -> bool {
+        self.0.get() % 8 == 0
+    }
+
+    /// Returns the maximum area size available for allocation.
+    #[expect(clippy::missing_panics_doc)]
+    #[inline]
+    #[must_use]
+    pub const fn max_area_size() -> u64 {
+        const { *AREA_SIZES.last().unwrap() }
+    }
+
+    /// Returns the minimum area size available for allocation.
+    #[expect(clippy::missing_panics_doc)]
+    #[inline]
+    #[must_use]
+    pub const fn min_area_size() -> u64 {
+        const { *AREA_SIZES.first().unwrap() }
+    }
 
-pub type FreeLists = [Option<LinearAddress>; NUM_AREA_SIZES];
+    /// Returns the number of different area sizes available.
+    #[inline]
+    #[must_use]
+    pub const fn num_area_sizes() -> usize {
+        const { AREA_SIZES.len() }
+    }
+
+    /// Returns a reference to the inner `NonZeroU64`
+    #[inline]
+    #[must_use]
+    pub const fn as_nonzero(&self) -> &NonZeroU64 {
+        &self.0
+    }
+
+    /// Advances a `LinearAddress` by `n` bytes.
+    ///
+    /// Returns `None` if the result overflows a u64
+    /// Some(LinearAddress) otherwise
+    ///
+    #[expect(clippy::missing_panics_doc)]
+    #[inline]
+    #[must_use]
+    pub const fn advance(self, n: u64) -> Option<Self> {
+        match self.0.get().checked_add(n) {
+            // overflowed
+            None => None,
+
+            // It is impossible to add a non-zero positive number to a u64 and get 0 without
+            // overflowing, so we don't check for that here, and panic instead.
+            Some(sum) => Some(LinearAddress(
+                NonZeroU64::new(sum).expect("sum cannot be zero"),
+            )),
+        }
+    }
```

Comment:
use [checked_add](https://doc.rust-lang.org/stable/std/num/type.NonZeroU64.html#method.checked_add) on the nonzero and not the u64. It will return a `Some(NonZero<_>)`
```suggestion
        match self.0.checked_add(n) {
            // overflowed
            None => None,
            Some(sum) => Some(LinearAddress(sum)),
        }
    }
```